calico_rkt/calico_rkt.py

```diff
@@ -14,387 +14,473 @@
-                              container_id=container_id,
-                              netns_path=netns_path,
-                              interface=interface,
-                              subnet=subnet)
-
-    _log.info('Finished configuring network interface')
-    return endpoint
-
-
-def _container_add(hostname, orchestrator_id, container_id, netns_path, interface, subnet):
-    """
-    Add a container to Calico networking
-    Return Endpoint object and newly allocated IP
-
-    :param hostname (str): Host for enndpoint allocation
-    :param orchestrator_id (str): Specifies orchestrator ('rkt')
-    :param container_id (str):
-    :param netns_path (str): namespace path
-    :param interface (str): iface to use
-    :param subnet (str): Subnet to allocate IP to
-    :rtype Endpoint: Endpoint created
-    """
-    # Allocate and Assign ip address through datastore_client
-    pool, ip = _assign_to_pool(subnet)
-
-    # Create Endpoint object
-    try:
-        ep = datastore_client.create_endpoint(HOSTNAME, ORCHESTRATOR_ID,
-                                              container_id, [ip])
-    except AddrFormatError:
-        _log.error("This node is not configured for IPv%d. Unassigning IP "
-                   "address %s then exiting." % ip.version, ip)
-        datastore_client.unassign_address(pool, ip)
-        sys.exit(1)
-
-    # Create the veth, move into the container namespace, add the IP and
-    # set up the default routes.
-    ep.mac = ep.provision_veth(Namespace(netns_path), interface)
-    datastore_client.set_endpoint(ep)
-    return ep
-
-
-def _container_remove(hostname, orchestrator_id, container_id):
-    """
-    Remove the indicated container on this host from Calico networking
+        :return: Response from the IPAM plugin.
+        """
+        # Find the correct plugin based on the given type.
+        plugin_path = self._find_ipam_plugin()
+        _log.debug("Using IPAM plugin at: %s", plugin_path)
+    
+        # Execute the plugin and return the result.
+        p = Popen(plugin_path, stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = p.communicate(json.dumps(self.network_config))
+        _log.debug("IPAM plugin output: \nstdout: %s\nstderr: %s", 
+                   stdout, stderr)
+        return stdout
+
+    def _create_endpoint(self, assigned_ip):
+        """Creates an endpoint in the Calico datastore with the client.
+
+        :param assigned_ip - IPAddress that has been already allocated
+        :return Calico endpoint object
+        """
+        try:
+            endpoint = self._client.create_endpoint(HOSTNAME,
+                                                    ORCHESTRATOR_ID,
+                                                    self.container_id,
+                                                    [assigned_ip])
+        except AddrFormatError:
+            _log.error("Unable to create endpoint. An invalid IP address was "
```

Comment:
I think this log is a little bit wrong.  `create_endpoint` also throws an `AddrFormatError` when an IPv6 address is provided, but IPv6 is not configured on the host (and vice-versa) even if the IPv6 address is valid.